### PR TITLE
feat: add core message components and store

### PR DESCRIPTION
**Pull request by elisealix22**
## What does this PR do?

This PR fills in the Messages component with some real data. It also introduces a Conversation component & initial XMTP store for state management. This should give us a good working ground to be able to work in parallel.

There is still A LOT to be done here so everything is behind the `messages` feature flag.

<details>
<summary>Screenshot</summary>

Screen Shot 2022-10-05 at 1 39 52 PM

</details>

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## How should this be tested?

- [x] Tap the messages nav button on `testnet`
- [x] Sign in with xmtp and ensure xmtp messages render in the messages component
- [x] Tap one and ensure it shows in the conversations component
- [x] Ensure the messages & conversation pages are not visible when the FF is disabled

### CODE REVIEW

**Review by bigint — CHANGES_REQUESTED** [review]

File: src/components/utils/hooks/useStreamMessages.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import { Conversation, Message, Stream } from '@xmtp/xmtp-js';
+import { useEffect, useState } from 'react';
+import { useXmtpStore } from 'src/store/xmtp';
+
+const useStreamMessages = (conversation: Conversation, onMessageCallback: () => void) => {
```
Comment (elisealix22) [inline src/components/utils/hooks/useStreamMessages.tsx:5]: @bhavya2611 I might be missing something, but I don't see this being used right now. Was this meant for the `Messages` component where we are streaming currently?

File: src/components/Conversation/index.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  if (!isFeatureEnabled('messages', currentProfile?.id)) {
+    return <Custom404 />;
+  }
+
+  return (
+    <GridLayout>
+      <MetaTags title={`Conversation • ${APP_NAME}`} />
+      <GridItemFour>
+        <Card className="h-[86vh] px-2 pt-3">
+          <div className="flex justify-between">
```
Comment (bigint) [inline src/components/Conversation/index.tsx:38]: **suggestion:** make sure to add `items-center`

File: src/components/Conversation/index.tsx [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            <div className="font-black text-lg">Messages</div>
```
Comment (bigint) [inline src/components/Conversation/index.tsx:39]:
```suggestion
            <div className="font-bold text-lg">Messages</div>
```

File: src/components/Conversation/index.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            <div>
+              <button className="text-xs border border-p-100 p-1 rounded">New Message</button>
```
Comment (bigint) [inline src/components/Conversation/index.tsx:41]: **suggestion:** we can use `Button` component with secondary outlined variant 🙇🏼

File: src/components/Messages/index.tsx [hunk 5]
```diff
@@ -1,13 +1,95 @@
...
+  profile: Profile;
+}
+
+const Messages: FC<Props> = () => {
+  const { data: signer } = useSigner();
+  const { address } = useAccount();
   const currentProfile = useAppStore((state) => state.currentProfile);
+  const [stream, setStream] = useState<Stream<Conversation>>();
+  const xmtpState = useXmtpStore((state) => state);
+  const { client, setClient, conversations, setConversations, messages, setMessages, setLoading } = xmtpState;
```
Comment (bigint) [inline src/components/Messages/index.tsx:25]: **question:** will it select only the used state or affect all the states?

https://github.com/pmndrs/zustand#selecting-multiple-state-slices

File: src/components/Messages/index.tsx [hunk 6]
```diff
@@ -1,13 +1,95 @@
...
+          messages.set(convo.peerAddress, newMessages);
+          setMessages(new Map(messages));
+          conversations.set(convo.peerAddress, convo);
+          setConversations(new Map(conversations));
+        }
+      }
+    };
+    listConversations();
+    streamConversations();
+    return () => {
```
Comment (bigint) [inline src/components/Messages/index.tsx:78]:
```suggestion
    listConversations();
    streamConversations();

    return () => {
```

suggested to leave new line before return 🚀

File: src/components/Messages/index.tsx [hunk 7]
```diff
@@ -1,13 +1,95 @@
...
 import { useAppStore } from 'src/store/app';
+import { useXmtpStore } from 'src/store/xmtp';
+import { useAccount, useSigner } from 'wagmi';
 
-const Messages: NextPage = () => {
+interface Props {
+  profile: Profile;
+}
+
+const Messages: FC<Props> = () => {
```
Comment (bigint) [inline src/components/Messages/index.tsx:19]: **suggestion:** `Props` are not used 🤔

**Review by bigint — CHANGES_REQUESTED** [review]
@elisealix22 @bhavya2611 left some comments 🙏🏼

File: src/components/utils/hooks/useStreamMessages.tsx (hunk 1 above)
Comment (bhavya2611) [reply src/components/utils/hooks/useStreamMessages.tsx:5]: Hey we will need this in near future, so having this makes sense

File: src/components/Conversation/index.tsx (hunk 4 above)
Comment (bhavya2611) [reply src/components/Conversation/index.tsx:41]: Styling is just dummy, for now, this is something we will be working on as per the designs. I suggest we ignore everything related to styling for now.

File: src/components/Conversation/index.tsx (hunk 3 above)
Comment (bhavya2611) [reply src/components/Conversation/index.tsx:39]: Styling is just dummy, for now, this is something we will be working on as per the designs. I suggest we ignore everything related to styling for now

File: src/components/Conversation/index.tsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            </div>
```
[the remark this answers is not in the document]
Comment (bhavya2611) [reply src/components/Conversation/index.tsx:42]: Styling is just dummy, for now, this is something we will be working on as per the designs. I suggest we ignore everything related to styling for now

File: src/components/Conversation/index.tsx [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+  if (!isFeatureEnabled('messages', currentProfile?.id)) {
+    return <Custom404 />;
+  }
+
+  return (
+    <GridLayout>
+      <MetaTags title={`Conversation • ${APP_NAME}`} />
+      <GridItemFour>
+        <Card className="h-[86vh] px-2 pt-3">
```
[the remark this answers is not in the document]
Comment (bhavya2611) [reply src/components/Conversation/index.tsx:37]: Styling is just dummy, for now, this is something we will be working on as per the designs. I suggest we ignore everything related to styling for now

File: src/components/Conversation/index.tsx [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+          <div className="flex justify-between">
+            <div className="font-black text-lg">Messages</div>
+            <div>
+              <button className="text-xs border border-p-100 p-1 rounded">New Message</button>
+            </div>
+          </div>
+          <div className="flex justify-between p-4">
+            <div className="text-xs">Lens profiles</div>
+            <div className="text-xs">All messages</div>
```
[the remark this answers is not in the document]
Comment (bhavya2611) [reply src/components/Conversation/index.tsx:46]: Styling is just dummy, for now, this is something we will be working on as per the designs. I suggest we ignore everything related to styling for now

File: src/components/Conversation/index.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (bhavya2611) [reply src/components/Conversation/index.tsx:37]: Styling is just dummy, for now, this is something we will be working on as per the designs. I suggest we ignore everything related to styling for now

File: src/components/Messages/message.tsx [hunk 11]
```diff
@@ -0,0 +1,80 @@
+import { Card } from '@components/UI/Card';
+import { GridItemEight, GridItemFour, GridLayout } from '@components/UI/GridLayout';
+import MetaTags from '@components/utils/MetaTags';
+import { Profile } from '@generated/types';
+import isFeatureEnabled from '@lib/isFeatureEnabled';
+import { Message } from '@xmtp/xmtp-js';
```
Comment (bigint) [inline src/components/Messages/message.tsx:6]: **nitpick:** conflict with a component name better alias the type 🤔

File: src/store/xmtp.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable no-unused-vars */
+import { Client, Conversation, Message } from '@xmtp/xmtp-js';
+import create from 'zustand';
+
+interface XmtpState {
```
Comment (bigint) [inline src/store/xmtp.ts:5]: **nitpick:** rename to `MessageState`

File: src/store/xmtp.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable no-unused-vars */
```
Comment (bigint) [inline src/store/xmtp.ts:1]: **suggestion:** better rename file to `message.ts`

**Review by bigint — CHANGES_REQUESTED** [review]

File: src/components/Messages/index.tsx [hunk 14]
```diff
@@ -1,13 +1,92 @@
...
+        })
+      ).then(() => {
+        setLoading(false);
+      });
+    }
+    const streamConversations = async () => {
+      const newStream = (await client?.conversations?.stream()) || [];
+      setStream(newStream);
+      for await (const convo of newStream) {
+        if (convo.peerAddress !== address) {
```
Comment (bigint) [inline src/components/Messages/index.tsx:63]: **nitpick:** use `currentProfile?.ownedBy`

File: src/components/Messages/index.tsx (hunk 5 above)
Comment (bigint) [reply src/components/Messages/index.tsx]: We can visit this later 🙏🏼

**Review by bigint — APPROVED** [review]
**LGTM** :shipit: 

Rad LGTM

**Pull request merged**